packages/commonwealth/client/scripts/views/pages/view_thread/edit_body.tsx

```diff
@@ -22,13 +23,50 @@ type EditBodyProps = {
 export const EditBody = (props: EditBodyProps) => {
 
   const { shouldRestoreEdits, savedEdits, thread, setIsEditing, title } = props;
-  const body = shouldRestoreEdits && savedEdits ? savedEdits : thread.body;
 
   const navigate = useCommonNavigate();
-  const [contentDelta, setContentDelta] = React.useState<DeltaStatic>(createDeltaFromText(body));
+
+  const threadBody = (shouldRestoreEdits && savedEdits) ? savedEdits : thread.body;
+  const body = parseDeltaString(threadBody)
+
+  const [contentDelta, setContentDelta] = React.useState<DeltaStatic>(body);
   const [saving, setSaving] = React.useState<boolean>(false);
-  
-  const editorValue = getTextFromDelta(contentDelta);
+
+  const cancel = async (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
+    e.preventDefault();
+
+    let cancelConfirmed = true;
+
+    if (JSON.stringify(body) !== JSON.stringify(contentDelta)) {
+      cancelConfirmed = window.confirm(
+        'Cancel editing? Changes will not be saved.'
+      );
+    }
+
+    if (cancelConfirmed) {
+      setIsEditing(false);
+      clearEditingLocalStorage(thread.id, ContentType.Thread);
+      navigate(`/discussion/${thread.id}`);
+    }
+  }
+
+  const save = async (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
+    e.preventDefault();
+
+    setSaving(true);
+
+    try {
+      await app.threads.edit(thread, JSON.stringify(contentDelta), title)
+      setIsEditing(false);
+      clearEditingLocalStorage(thread.id, ContentType.Thread);
+      navigate(`/discussion/${thread.id}`);
```

Comment:
this is not good. I mean, it works, but first of all, it is not a react way and second, it is a visible page blink on cancel or save. It should work in similar way as in comment (updatedCommentsCallback). The callback should update the thread state up in the tree and the page should rerender after state change, not with force reload

Reply:
Refactored to use the similar pattern.